File: code/design-patterns/creational/builder/python/builder.py
from abc import ABC, abstractmethod


class House:
    def __init__(self):
        self.foundation = None
        self.structure = None
        self.roof = None
        self.rooms = None
        self.backyard = None

    def __str__(self):
        return (
            f"House with foundation: {self.foundation}, structure: {self.structure}, "
            f"roof: {self.roof}, rooms: {self.rooms}, backyard: {self.backyard}"
        )


class HouseBuilder(ABC):
    @abstractmethod
    def build_foundation(self):
        pass

    @abstractmethod
    def build_structure(self):
        pass

    @abstractmethod
    def build_roof(self):
        pass

    @abstractmethod
    def build_rooms(self, number):
        pass

    @abstractmethod
    def build_backyard(self, type):
        pass

    @abstractmethod
    def build(self):
        pass


class WoodenHouseBuilder(HouseBuilder):
    def __init__(self):
        self.house = House()

    def build_foundation(self):
        self.house.foundation = "Wooden Foundation"
        return self

    def build_structure(self):
        self.house.structure = "Wooden Structure"
        return self

    def build_roof(self):
        self.house.roof = "Wooden Roof"
        return self

    def build_rooms(self, number):
        self.house.rooms = f"{number} Wooden Rooms"
        return self

    def build_backyard(self, type):
        self.house.backyard = f"Wooden {type} Backyard"
        return self

    def build(self):
        return self.house


class BrickHouseBuilder(HouseBuilder):
    def __init__(self):
        self.house = House()

    def build_foundation(self):
        self.house.foundation = "Concrete Foundation"
        return self

    def build_structure(self):
        self.house.structure = "Brick Structure"
        return self

    def build_roof(self):
        self.house.roof = "Concrete Roof"
        return self

    def build_rooms(self, number):
        self.house.rooms = f"{number} Brick Rooms"
        return self

    def build_backyard(self, type):
        self.house.backyard = f"Brick {type} Backyard"
        return self

    def build(self):
        return self.house


class Director:
    def __init__(self, builder):
        self._builder = builder

    def construct_simple_house(self):
        self._builder.build_foundation().build_structure().build_roof().build_rooms(2)
        return self._builder.build()

    def construct_luxury_house(self):
        self._builder.build_foundation().build_structure().build_roof().build_rooms(
            5
        ).build_backyard("Luxury")
        return self._builder.build()

File: code/design-patterns/creational/builder/go/builder.go
package builder

import (
	"fmt"
)

type House struct {
	Foundation string
	Structure  string
	Roof       string
	Rooms      string
	Backyard   string
}

func (h House) String() string {
	return fmt.Sprintf("House with foundation: %s, structure: %s, roof: %s, rooms: %s, backyard: %s",
		h.Foundation, h.Structure, h.Roof, h.Rooms, h.Backyard)
}

type HouseBuilder interface {
	BuildFoundation() HouseBuilder
	BuildStructure() HouseBuilder
	BuildRoof() HouseBuilder
	BuildRooms(number int) HouseBuilder
	BuildBackyard(typ string) HouseBuilder
	Build() House
}

type WoodenHouseBuilder struct {
	house House
}

func (b *WoodenHouseBuilder) BuildFoundation() HouseBuilder {
	b.house.Foundation = "Wooden Foundation"
	return b
}

func (b *WoodenHouseBuilder) BuildStructure() HouseBuilder {
	b.house.Structure = "Wooden Structure"
	return b
}

func (b *WoodenHouseBuilder) BuildRoof() HouseBuilder {
	b.house.Roof = "Wooden Roof"
	return b
}

func (b *WoodenHouseBuilder) BuildRooms(number int) HouseBuilder {
	b.house.Rooms = fmt.Sprintf("%d Wooden Rooms", number)
	return b
}

func (b *WoodenHouseBuilder) BuildBackyard(typ string) HouseBuilder {
	b.house.Backyard = fmt.Sprintf("Wooden %s Backyard", typ)
	return b
}

func (b *WoodenHouseBuilder) Build() House {
	return b.house
}

type BrickHouseBuilder struct {
	house House
}

func (b *BrickHouseBuilder) BuildFoundation() HouseBuilder {
	b.house.Foundation = "Concrete Foundation"
	return b
}

func (b *BrickHouseBuilder) BuildStructure() HouseBuilder {
	b.house.Structure = "Brick Structure"
	return b
}

func (b *BrickHouseBuilder) BuildRoof() HouseBuilder {
	b.house.Roof = "Concrete Roof"
	return b
}

func (b *BrickHouseBuilder) BuildRooms(number int) HouseBuilder {
	b.house.Rooms = fmt.Sprintf("%d Brick Rooms", number)
	return b
}

func (b *BrickHouseBuilder) BuildBackyard(typ string) HouseBuilder {
	b.house.Backyard = fmt.Sprintf("Brick %s Backyard", typ)
	return b
}

func (b *BrickHouseBuilder) Build() House {
	return b.house
}

type Director struct {
	builder HouseBuilder
}

func (d *Director) ConstructSimpleHouse() House {
	return d.builder.BuildFoundation().BuildStructure().BuildRoof().BuildRooms(2).Build()
}

func (d *Director) ConstructLuxuryHouse() House {
	return d.builder.BuildFoundation().BuildStructure().BuildRoof().BuildRooms(5).BuildBackyard("Luxury").Build()
}

func (d *Director) ConstructHouseWith3Bedrooms() House {
	return d.builder.BuildFoundation().BuildStructure().BuildRoof().BuildRooms(3).BuildBackyard("Standard").Build()
}
